design_documents/contract_verification.md

```diff
@@ -0,0 +1,131 @@
+# Cairo Contract Verification via Block Explorers
+
+## Context
+
+Cairo smart contracts deployed to Starknet are only visible as a Cairo Bytecode and their's ABIs, which is difficult to decipher and understand. To increase readability of a smart contract, it's source code can also be made public, however in order to ensure trustlessness, the source code vs resulting Cairo Bytecode verification can be performed. Such verification and its results can be performed and made public by popular Blockchain Explorers.
+
+## Goal
+
+This proposal includes an extension to `sncast` utility enabling a contract owner to perform contract verification against a selected Blockchain Explorer API. We propose to create a first, reference implementation for the Voyager APIs.
+
+## Proposed Solution
+
+We propose to design a dedicated `verify` command for the `sncast` tool, and add a mechanism whereby this command can be implemented for various Blockchain Explorers. We define a "generic" contract verification interface and propose to implement the interface by "adapters" specific to respective explorers. 
+
+### `sncast` utility command - `verify`
+
+Command name: `verify`
+
+The `verify` command will perform following actions:
+- select the verifiation logic to use (based on `--verifier` parameter)
+- pick the selected scarb workspace source code from local filesystem (based on `Scarb.toml` file)
+- for the selected verifier it will upload the Scarb workspace contracts source code to the verifier's API
+- call the verifier's API to trigger verification of the uploaded source code
+- wait for results of the verification API call
+- respond to the user with the results 
+
+#### Parameters
+
+#### `--contract-name`
+
+Required.
+Name of the contract to be submitted for verification.
+
+#### `--contract-address`
+
+Required.
+Address of the contract to be submitted for verification.
+
+#### `--verifier <VERIFIER NAME>`
+
+Optional.
+Specifies the Blockchan Explorer to verify with.  
+Default: `voyager`
```

Comment:
I'd prefer we don't have a default defined in the document itself

Reply:
So you want to make this argument a required one ?

Reply:
I think it might be good for the user to explicitly select the block explorer. cc @Arcticae @THenry14 what do you think?

Reply:
Yes i think that is reasonable, since there might be other available ones in the future and we don't want to be biased :)